### PR TITLE
 [chinese text normalization] PR for chinese TN part in text_normalization 

### PR DESCRIPTION
**Pull request by mzxcpp**
# What does this PR do ?

PR for chinese TN part in text_normalization 

**Collection**: NeMo/norm_text_processing/text_normalization

Chinese Text Normalization
1. How To Use
python normalize.py --lang "zh" --text "text to be normalized"
2. TN Pipeline
There are 3 components in TN pipeline:

pre-processing (before tagger)
non-standard word normalization
post-processing (after verbalizer)
2.1 Pre-Processing
Char Width Conversion (全角 -> 半角)
苹果ＣＥＯ宣布发布新ＩＰＨＯＮＥ -> 苹果CEO宣布发布新IPHONE
他说：“我们已经吃过了！”。 -> 他说:"我们已经吃过了!".
covers English letters, digits, punctuations and some symbols
the complete mapping table data/char/fullwidth_to_halfwidth.tsv
Blacklist (Removal)
Sometime you may want to remove certain things like interjections/fillers "啊", "呃" etc

呃这个呃啊我不知道 -> 这个我不知道
customizable via data/blacklist/interjections.tsv
2.2 Non-Standard-Words(NSW) normalization
Numbers
  共465篇，约315万字 -> 共四百六十五篇，约三百一十五万字
  共计6.42万人 -> 共计六点四二万人
  同比升高0.6个百分点 -> 同比升高零点六个百分点
Fraction
  总量的1/5以上 -> 总量的五分之一以上
  相当于头发丝的1/16 -> 相当于头发丝的十六分之一
  3/2是一个假分数 -> 二分之三是一个假分数
Percentage
  同比增长6.3% -> 同比增长百分之六点三
  增幅0.4% -> 增幅百分之零点四
Date
  2002/01/28 -> 二零零二年一月二十八日
  2002-01-28 -> 二零零二年一月二十八日
  2002.01.28 -> 二零零二年一月二十八日
  2002/01 -> 二零零二年一月
Time
  8月16号12:00之前 -> 八月十六号十二点之前
  我是5:02开始的 -> 我是五点零二分开始的
  于5:35:36发射 -> 于五点三十五分三十六秒发射
  8:00 a.m. 准时开会 -> 八点 a m 准时开会
Math
  比分定格在78:96 -> 比分定格在七十八比九十六
  计算-2的绝对值是2 -> 计算负二的绝对值是二
  ±2的平方都是4 -> 正负二的平方都是四
Money
  价格是￥13.5 -> 价格是十三点五元
  价格是$13.5 -> 价格是十三点五美元
  价格是A$13.5 -> 价格是十三点五澳元
  价格是HKD13.5 -> 价格是十三点五港元
Measure
  重达25kg -> 二十五千克
  最高气温38°C -> 最高气温三十八摄氏度
  实际面积120m² -> 实际面积一百二十平方米
  渲染速度10ms一帧 -> 渲染速度十毫秒一帧
Number series (phone, mobile numbers)
  可以打我手机13501234567 -> 可以打我手机一三五零一二三四五六七
  可以拨打12306来咨询 -> 可以拨打一二三零六来咨询
Erhua(儿化音) Removal
  这儿有只鸟儿 -> 这有只鸟
  这事儿好办 -> 这事好办
  我儿子喜欢这地儿 -> 我儿子喜欢这地
erhua whitelist is customizable via data/erhua/whitelist.tsv
Whitelist(Replacement)
a set of user-defined hard mapping, i.e. exact-string matching & replacement

CEO -> C E O
GPU -> G P U
O2O -> O to O
B2B -> B to B
customizable via data/whitelist/default.tsv
2.3 Post-Processing
Punctuation Removal
If enabled, punctuations are removed.

Uppercase or Lowercase Conversion
If enabled, English letters are converted to uppercases / lowercases

Out-Of-Vocabulary(OOV) Tagger
If enabled, OOV chars are tagged with <oov> and </oov>, e.g.:

我们안녕 -> 我们<oov>안</oov><oov>녕</oov>
雪の花 -> 雪<oov>の</oov>花
default charset (national standard) [通用规范汉字表](https://zh.wikipedia.org/wiki/%E9%80%9A%E7%94%A8%E8%A7%84%E8%8C%83%E6%B1%89%E5%AD%97%E8%A1%A8)
you can extend charset via data/char/charset_extension.tsv
3. Credits
Author: Zhenxiang MA @ Tsinghua University

Advisors: [SpeechColab](https://github.com/SpeechColab) organization

The authors of this work would like to thank:

The authors of foundational libraries like OpenFst & Pynini
NeMo team and NeMo open-source community

### CODE REVIEW

**Review by mzxcpp — COMMENTED** [review]
add chinese tn part

**Comment by dophist** [comment]
This is the resulting work from https://github.com/NVIDIA/NeMo/issues/4543

README is pasted below to save reviewers' time.

# Chinese Text Normalization

 ## 1. How To Use
 ```
 python normalize.py --lang "zh" --text "text to be normalized"
 ```

 ## 2. TN Pipeline
 There are 3 components in TN pipeline:
 * pre-processing (before tagger)
 * non-standard word normalization
 * post-processing (after verbalizer)

 ### 2.1 Pre-Processing
 #### Char Width Conversion (全角 -> 半角)
 ```
 苹果ＣＥＯ宣布发布新ＩＰＨＯＮＥ -> 苹果CEO宣布发布新IPHONE
 他说：“我们已经吃过了！”。 -> 他说:"我们已经吃过了!".
 ```
 * covers English letters, digits, punctuations and some symbols
 * the complete mapping table `data/char/fullwidth_to_halfwidth.tsv`

 #### Blacklist (Removal)
 Sometime you may want to remove certain things like interjections/fillers "啊", "呃" etc
 ```
 呃这个呃啊我不知道 -> 这个我不知道
 ```
 * customizable via `data/blacklist/interjections.tsv`


 ### 2.2 Non-Standard-Words(NSW) normalization
 #### Numbers
 ```
 共465篇，约315万字 -> 共四百六十五篇，约三百一十五万字
 共计6.42万人 -> 共计六点四二万人
 同比升高0.6个百分点 -> 同比升高零点六个百分点
 ```

 #### Fraction
 ```
 总量的1/5以上 -> 总量的五分之一以上
 相当于头发丝的1/16 -> 相当于头发丝的十六分之一
 3/2是一个假分数 -> 二分之三是一个假分数
 ```

 #### Percentage
 ```
 同比增长6.3% -> 同比增长百分之六点三
 增幅0.4% -> 增幅百分之零点四
 ```

 #### Date
 ```
 2002/01/28 -> 二零零二年一月二十八日
 2002-01-28 -> 二零零二年一月二十八日
 2002.01.28 -> 二零零二年一月二十八日
 2002/01 -> 二零零二年一月
 ```

 #### Time
 ```
 8月16号12:00之前 -> 八月十六号十二点之前
 我是5:02开始的 -> 我是五点零二分开始的
 于5:35:36发射 -> 于五点三十五分三十六秒发射
 8:00 a.m. 准时开会 -> 八点 a m 准时开会
 ```

 #### Math
 ```
 比分定格在78:96 -> 比分定格在七十八比九十六
 计算-2的绝对值是2 -> 计算负二的绝对值是二
 ±2的平方都是4 -> 正负二的平方都是四
 ```

 #### Money
 ```
 价格是￥13.5 -> 价格是十三点五元
 价格是$13.5 -> 价格是十三点五美元
 价格是A$13.5 -> 价格是十三点五澳元
 价格是HKD13.5 -> 价格是十三点五港元
 ```

 #### Measure
 ```
 重达25kg -> 二十五千克
 最高气温38°C -> 最高气温三十八摄氏度
 实际面积120m² -> 实际面积一百二十平方米
 渲染速度10ms一帧 -> 渲染速度十毫秒一帧
 ```

 #### Number series (phone, mobile numbers)
 ```
 可以打我手机13501234567 -> 可以打我手机一三五零一二三四五六七
 可以拨打12306来咨询 -> 可以拨打一二三零六来咨询
 ```

 #### Erhua(儿化音) Removal
 ```
 这儿有只鸟儿 -> 这有只鸟
 这事儿好办 -> 这事好办
 我儿子喜欢这地儿 -> 我儿子喜欢这地
 ```
 * erhua whitelist is customizable via `data/erhua/whitelist.tsv`

 #### Whitelist(Replacement)
 a set of user-defined hard mapping, i.e. exact-string matching & replacement
 ```
 CEO -> C E O
 GPU -> G P U
 O2O -> O to O
 B2B -> B to B
 ```
 * customizable via `data/whitelist/default.tsv`

 ### 2.3 Post-Processing
 #### Punctuation Removal
 If enabled, punctuations are removed.

 #### Uppercase or Lowercase Conversion
 If enabled, English letters are converted to uppercases / lowercases

 #### Out-Of-Vocabulary(OOV) Tagger
 If enabled, OOV chars are tagged with `<oov>` and `</oov>`, e.g.:
 ```
 我们안녕 -> 我们<oov>안</oov><oov>녕</oov>
 雪の花 -> 雪<oov>の</oov>花
 ```
 * default charset (national standard) [通用规范汉字表](https://zh.wikipedia.org/wiki/通用规范汉字表)
 * you can extend charset via `data/char/charset_extension.tsv`

 ## 3. Credits
 Author: Zhenxiang MA @ Tsinghua University

 Advisors: [SpeechColab](https://github.com/SpeechColab) organization

 The authors of this work would like to thank:
 * The authors of foundational libraries like OpenFst & Pynini
 * NeMo team and NeMo open-source community

File: nemo_text_processing/text_normalization/token_parser.py [hunk 1]
```diff
@@ -156,7 +156,8 @@ def parse_string_value(self) -> str:
 
         Returns parsed string value
         """
-        assert self.char not in string.whitespace and self.char != EOS
+        #assert self.char not in string.whitespace and self.char != EOS
+        assert self.char != EOS
```
Comment (ekmb) [inline nemo_text_processing/text_normalization/token_parser.py:160]: Why this change is needed?

File: nemo_text_processing/text_normalization/zh/case.tsv [hunk 2]
```diff
@@ -0,0 +1,47 @@
+苹果ＣＥＯ宣布发布新ＩＰＨＯＮＥ
```
Comment (ekmb) [inline nemo_text_processing/text_normalization/zh/case.tsv:1]: What is the role of this file?

**Review by ekmb — COMMENTED** [review]
Thank you for your contribution, @mzxcpp ! Could you please add test cases similar to [https://github.com/NVIDIA/NeMo/tree/main/tests/nemo_text_processing](https://github.com/NVIDIA/NeMo/tree/main/tests/nemo_text_processing)

File: nemo_text_processing/text_normalization/zh/case.tsv (hunk 2 above)
Comment (mzxcpp) [reply nemo_text_processing/text_normalization/zh/case.tsv:1]: this is the test cases supported now.This case.tav cover all the different cases now.

Comment (ekmb) [reply nemo_text_processing/text_normalization/zh/case.tsv:1]: Could please reformat them like [here](https://github.com/NVIDIA/NeMo/tree/main/tests/nemo_text_processing)?

File: nemo_text_processing/text_normalization/token_parser.py (hunk 1 above)
Comment (mzxcpp) [reply nemo_text_processing/text_normalization/token_parser.py:160]: In Chinese, not every Chinese character has a space between them, and the spaces in a sentence have a special meaning that needs to be preserved.In this case,I have extended this part of the code

File: nemo_text_processing/text_normalization/zh/README.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
+# Chinese Text Normalization
+
+## 1. How To Use
+```
+python normalize.py --lang "zh" --text "text to be normalized"
```
Comment (dophist) [inline nemo_text_processing/text_normalization/zh/README.md:5]: --lang  ->  --language

File: nemo_text_processing/text_normalization/zh/data/char/charset_illegal_tags.tsv [hunk 4]
```diff
@@ -0,0 +1 @@
+<	>
```
Comment (dophist) [inline nemo_text_processing/text_normalization/zh/data/char/charset_illegal_tags.tsv:1]: <oov> & </oov>

Comment (dophist) [reply nemo_text_processing/text_normalization/zh/data/char/charset_illegal_tags.tsv:1]: "charset_illegal_tags.tsv" is out-dated resource

File: nemo_text_processing/text_normalization/zh/utils.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        additional labels
+    """
+    res = []
+    for label in labels:
+        if len(label) > 1:
+            if label[0][-1] == "." and label[1][-1] != ".":
+                res.append([label[0], label[1] + "."] + label[2:])
+    return res
+
+def inverse_chr_sep(line):
```
Comment (dophist) [inline nemo_text_processing/text_normalization/zh/utils.py:70]: Is this function still used ?

File: nemo_text_processing/text_normalization/zh/data/time/hour.tsv [hunk 6]
```diff
@@ -0,0 +1,24 @@
+1
```
Comment (dophist) [inline nemo_text_processing/text_normalization/zh/data/time/hour.tsv:1]: these resources are out-dated, have been renamed to clock

File: nemo_text_processing/text_normalization/token_parser.py (hunk 1 above)
Comment (dophist) [inline nemo_text_processing/text_normalization/token_parser.py:160]: @ekmb Evelina, the relaxation of this assertion is due to the fact that modern Chinese is frequently mixed with English phrases. E.g.:

“听首Hey Jude" (in English, it means "play Hey Jude")

we want those English segment to pass through the TN pipeline intactly.   This is not a Chinese-specific issue, any char-based language that can be mixed with English phrases, will have this problem.   So our choices could be: 1) honor space as legit value, or 2) use a specific tag to represent space explicitly, and restore these spaces in verbalizer.

Current code uses option one by relaxing the assertion condition.  We could switch to option two, but eventually NeMo will have other languages with similar scenarios like what we have now.

@mzxcpp Zhenxiang, please follow up this issue, and adapt the code accordingly based on Evelina's suggestion.

**Comment by BuyuanCui** [comment]
Thank you for the hard work! I was browsing the grammar and did some test run myself. This art piece is amazing. Just a few things which I would like to mention.
1. The grammar did not cover traditional format for Mandarin character. My main focus has been on ITN grammar so far. At least for ITN, we do try to make sure that our grammar can also process traditional format.
2. Grammar Money: there is unnecessary space between number and symbol.
For example, I received an output '一百 元', where in written format, the space between '百' and '元' should not exist.
3. Grammar Date: This is a similar issue with Money where there are spaces between year, month, and day.
For example, in the output '二零零二年 一月 二十八日', we may not the space in the output. These are some minor stuff. 
4. Grammar Date2: wrong output. For input '1998年', the correct output is '一九九八年', where the tested output is '一千九百九十八年'. The result shows that the number is considered as cardinals where in Mandarin, the numbers for years should be recognized one-by-one (eg., units' digit: '1998year' should be 'one nine nine eight -year', but the output of the grammar is 'one thousand ninety eight year'). One possible solution for this is to simply change the weight for Cardinal grammar and Date grammar.
5. Grammar Math: wrong output. For input '1-1=0', the correct output is '一减一等于零', whereas the output is '一负一等于零'. The sing is realized as 'negative' sign instead of 'minus' sign.
6. Grammar Clock (e.g., Time in our NeMo). Ideally, the grammar should be able to process AM and PM. That is, when the input includes AM or PM, it should be processed as '上午' and '下午‘. It also does not cover time zones. I'm sure this won't be a challenge. 
7. There are typos in data file. In <tests/nemo_text_processing/zh/data_text_normalization/test_cases_fraction.txt>, the matchings between inputs and outputs are wrong. The input '2/3' is matched with '二分之三', where it has to be '三分之二' and '3/2' is matched to '三分之二'; the correct format is '二分之三'. 
Just some issues I found. I think it is, overall, well done.

**Comment by mzxcpp** [comment]
thank you for your advice.I have just fixed some problems yesterday and i will go through it again to try my best to cover all cases.

File: nemo_text_processing/text_normalization/zh/taggers/whitelist.py [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    '''
+        ATM  -> tokens { whitelist: "ATM" } 
+    '''
+    def __init__(self, deterministic: bool = True, lm: bool = False):
+        super().__init__(name="whitelist", kind="classify", deterministic=deterministic)
+
+        whitelist = pynini.string_file(get_abs_path("data/whitelist/default.tsv"))
+
+        graph = (
+            pynutil.insert("whitelist: \"")
```
Comment (ekmb) [inline nemo_text_processing/text_normalization/zh/taggers/whitelist.py:16]: Use `name:` field as in https://github.com/NVIDIA/NeMo/blob/main/nemo_text_processing/text_normalization/en/taggers/whitelist.py#L129

File: nemo_text_processing/text_normalization/zh/utils.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
+# Copyright (c) 2021, NVIDIA CORPORATION.  All rights reserved.
```
Comment (ekmb) [inline nemo_text_processing/text_normalization/zh/utils.py:1]: outdated header

File: nemo_text_processing/text_normalization/zh/verbalizers/char.py [hunk 9]
```diff
@@ -0,0 +1,14 @@
+import pynini
+from nemo_text_processing.text_normalization.zh.graph_utils import NEMO_CHAR, NEMO_SIGMA, GraphFst,NEMO_NOT_SPACE,NEMO_NOT_QUOTE
+from pynini.lib import pynutil
+from nemo_text_processing.text_normalization.zh.utils import get_abs_path
+
+class Char(GraphFst):
```
Comment (ekmb) [inline nemo_text_processing/text_normalization/zh/verbalizers/char.py:6]: Use `name:`  field instead of `char` like in https://github.com/mzxcpp/NeMo/blob/main/nemo_text_processing/text_normalization/en/taggers/word.py#L90

File: nemo_text_processing/text_normalization/zh/verbalizers/clock.py [hunk 10]
```diff
@@ -0,0 +1,54 @@
+import pynini
+from nemo_text_processing.text_normalization.zh.graph_utils import NEMO_SIGMA, GraphFst, insert_space,NEMO_DIGIT,NEMO_NOT_QUOTE, delete_space
+from nemo_text_processing.text_normalization.zh.utils import get_abs_path,UNIT_1e01
+from pynini.lib import pynutil
+
+class Clock(GraphFst):
```
Comment (ekmb) [inline nemo_text_processing/text_normalization/zh/verbalizers/clock.py:6]: Clock -> Time

File: nemo_text_processing/text_normalization/zh/taggers/erhua.py [hunk 11]
```diff
@@ -0,0 +1,23 @@
+import pynini
+from nemo_text_processing.text_normalization.zh.graph_utils import NEMO_SIGMA, GraphFst,NEMO_NOT_SPACE
+from pynini.lib import pynutil
+from nemo_text_processing.text_normalization.zh.utils import get_abs_path,load_labels
+
+class Erhua(GraphFst):
```
Comment (ekmb) [inline nemo_text_processing/text_normalization/zh/taggers/erhua.py:6]: Could this class be merged with the whitelist class?

File: tests/nemo_text_processing/zh/test_date.py [hunk 12]
```diff
@@ -0,0 +1,30 @@
+# Copyright (c) 2021, NVIDIA CORPORATION.  All rights reserved.
```
Comment (ekmb) [inline tests/nemo_text_processing/zh/test_date.py:1]: outdated headers in this folder

File: tests/nemo_text_processing/zh/data_text_normalization/test_cases_char.txt [hunk 13]
```diff
@@ -0,0 +1,2 @@
+你好~你好
```
Comment (ekmb) [inline tests/nemo_text_processing/zh/data_text_normalization/test_cases_char.txt:1]: Thank you for adding the test cases!

**Review by ekmb — CHANGES_REQUESTED** [review]
Thank you for the great PR!
Could you please update class names and field names to follow [Sparawhak proto (a framework for deployment)](https://github.com/yzhang123/sparrowhawk/blob/test/src/proto/semiotic_classes.proto). We can then export the grammars using https://github.com/NVIDIA/NeMo/blob/main/tools/text_processing_deployment/pynini_export.py#L76

**Comment by ekmb** [comment]
run `python setup.py style --fix` from `NeMo/` directory to fix formatting issues to pass CI tests

**Comment by mzxcpp** [comment]
> Thank you for the hard work! I was browsing the grammar and did some test run myself. This art piece is amazing. Just a few things which I would like to mention.
> 
> 1. The grammar did not cover traditional format for Mandarin character. My main focus has been on ITN grammar so far. At least for ITN, we do try to make sure that our grammar can also process traditional format.
> 2. Grammar Money: there is unnecessary space between number and symbol.
>    For example, I received an output '一百 元', where in written format, the space between '百' and '元' should not exist.
> 3. Grammar Date: This is a similar issue with Money where there are spaces between year, month, and day.
>    For example, in the output '二零零二年 一月 二十八日', we may not the space in the output. These are some minor stuff.
> 4. Grammar Date2: wrong output. For input '1998年', the correct output is '一九九八年', where the tested output is '一千九百九十八年'. The result shows that the number is considered as cardinals where in Mandarin, the numbers for years should be recognized one-by-one (eg., units' digit: '1998year' should be 'one nine nine eight -year', but the output of the grammar is 'one thousand ninety eight year'). One possible solution for this is to simply change the weight for Cardinal grammar and Date grammar.
> 5. Grammar Math: wrong output. For input '1-1=0', the correct output is '一减一等于零', whereas the output is '一负一等于零'. The sing is realized as 'negative' sign instead of 'minus' sign.
> 6. Grammar Clock (e.g., Time in our NeMo). Ideally, the grammar should be able to process AM and PM. That is, when the input includes AM or PM, it should be processed as '上午' and '下午‘. It also does not cover time zones. I'm sure this won't be a challenge.
> 7. There are typos in data file. In <tests/nemo_text_processing/zh/data_text_normalization/test_cases_fraction.txt>, the matchings between inputs and outputs are wrong. The input '2/3' is matched with '二分之三', where it has to be '三分之二' and '3/2' is matched to '三分之二'; the correct format is '二分之三'.
>    Just some issues I found. I think it is, overall, well done.

in chinese,like “98年” has two spoken forms as "九八年" or ”九十八年" in different cases and they are both common in daliy lifes, in this case, i set "xx年后“"xx年来" or just "xx年” ,with character in year_whitelist behind "年" to cardinal forms, while "xx年是" with character not in year_whitelist into number string.I make this rule to adapt to detailed chinese sentences and it act better than just numberstrig with “年”.In this case ,only just "1998年" will go wrong and other detailed cases just like "1998年，我出生了" will get the expected answer.

File: nemo_text_processing/text_normalization/zh/taggers/preprocessor.py [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+
+class PreProcessor(GraphFst):
+    '''
+        Preprocessing of TN:
+            1. interjections removal such as '啊, 呃'
+            2. fullwidth -> halfwidth char conversion
+    '''
+    def __init__(self,
+        remove_interjections: bool = True,
+        fullwidth_to_halfwidth: bool = True,
```
[the remark this answers is not in the document]
Comment (mzxcpp) [reply nemo_text_processing/text_normalization/zh/taggers/preprocessor.py:14]: the injection should be removed in asr-score tasks, while in other filed,like tts,injection needn't be removed.Just use it to suitable for different users.

**Comment by mzxcpp** [comment]
I have gone through all lines again.@ekmb

**Comment by dophist** [comment]
@mzxcpp Zhenxiang, I don't know how you incorporated so many irrelevant changes into this PR, which makes it impossible to review and dangerous to merge. 

Your work should be almost entirely restricted inside `zh/` directory, except an assertion relaxation relating to whitespace as I recall.  Please fix this PR, if this PR is totally messed up beyond your git skills, you may even consider open a new one, with issues from previous reviews addressed.

**Comment by mzxcpp** [comment]
> incorporated

I think I should update my branch and I have just update it many times,is it needless?I will try a new pr then.

**Comment by mzxcpp** [comment]
sorry for not familiar with git.I will create a new pr soon,and this pr close.
